Services/Storage/DocumentDb/DocumentDbWrapper.cs

```diff
@@ -0,0 +1,276 @@
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Error while getting DocumentDb database", e);
+                throw;
+            }
+        }
+
+        private async Task CreateDatabaseAsync(DocumentClient client, string dbName)
+        {
+            try
+            {
+                this.log.Info("Creating DocumentDb database",
+                    () => new { dbName });
+                var db = new Database { Id = dbName };
+                await client.CreateDatabaseAsync(db);
+            }
+            catch (DocumentClientException e) when (e.StatusCode == HttpStatusCode.Conflict)
+            {
+                this.log.Warn("Another process already created the database", () => new { dbName });
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Error while creating DocumentDb database", () => new { dbName, e });
+                throw;
+            }
+        }
+        private async Task CreateCollectionIfNotExistsAsync(
+            IDocumentClient client,
+            RequestOptions options,
+            string dbName,
+            string collName)
+        {
+            try
+            {
+                var uri = $"/dbs/{dbName}/colls/{collName}";
+                await client.ReadDocumentCollectionAsync(uri, options);
+            }
+            catch (DocumentClientException e) when (e.StatusCode == HttpStatusCode.NotFound)
+            {
+                await this.CreateCollectionAsync(client, dbName, collName, options);
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Error while getting DocumentDb collection", e);
+                throw;
+            }
+        }
+
+        private async Task CreateCollectionAsync(
+            IDocumentClient client,
+            string dbName,
+            string collName,
+            RequestOptions options)
+        {
+            try
+            {
+                this.log.Info("Creating DocumentDb collection",
+                    () => new { dbName, collName });
+                var coll = new DocumentCollection { Id = collName };
+
+                var index = Index.Range(DataType.String, -1);
+                var indexing = new IndexingPolicy(index) { IndexingMode = IndexingMode.Consistent };
+                coll.IndexingPolicy = indexing;
+                var dbUri = "/dbs/" + dbName;
+                await client.CreateDocumentCollectionAsync(dbUri, coll, options);
+            }
+            catch (DocumentClientException e)
+            {
+                if (e.StatusCode == HttpStatusCode.Conflict)
+                {
+                    this.log.Warn("Another process already created the collection",
+                        () => new { dbName, collName });
+                }
+
+                this.log.Error("Error while creating DocumentDb collection",
+                    () => new { dbName, collName, e });
```

Comment:
throw `ExternalDependencyException` here so that the caller knows something went wrong and it should retry

Reply:
Done.